File: C/4_datatypes/solution/b/derived.c
#include <stdio.h>
#include <stdlib.h>

#include <mpi.h>

const int MATRIX_DIM = 10;

#define INITIALIZE_MATRIX(matrix, size, expr) \
    for (int i = 0; i < size; i++) \
        for( int j = 0; j < size; j++) \
            matrix[i * size + j] = expr;

void print_matrix(int* matrix, int size)
{
    for (int i = 0; i < size; i++)
    {
        for( int j = 0; j < size; j++)
        {
            printf("%3i ", matrix[i*size+j]);
        }
        printf("\n");
    }
}

void print_typeinfo(MPI_Datatype type, const char* typename)
{
    int size;
    MPI_Aint lb, extent;
    MPI_Type_size(type, &size);
    MPI_Type_get_extent(type, &lb, &extent);
    printf("%s: size=%d, lb=%ld, extent=%ld\n", typename, size, lb, extent);
}

int main(int argc, char** argv)
{
    int my_rank, num_ranks, tag = 0;
    MPI_Datatype matrixtype, vectortype;

    int* matrix = calloc(MATRIX_DIM * MATRIX_DIM, sizeof(int));

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);
    MPI_Comm_size(MPI_COMM_WORLD, &num_ranks);

    if (my_rank == 0)
    {
        INITIALIZE_MATRIX(matrix, MATRIX_DIM, i*100+j);
    }
    else
    {
        INITIALIZE_MATRIX(matrix, MATRIX_DIM, 0);
    }

    if (my_rank == 0)
    {
        int count = MATRIX_DIM, blocklength = 1, stride = MATRIX_DIM;
        MPI_Type_vector(count, blocklength, stride, MPI_INT, &vectortype);

        int intsize;
        MPI_Type_size(MPI_INT, &intsize);
        int blocklengths[MATRIX_DIM];
        MPI_Aint displs[MATRIX_DIM];
        for (int i = 0; i < MATRIX_DIM; ++i)
        {
            blocklengths[i] = 1;
            displs[i] = i * intsize;
        }
        MPI_Type_create_hindexed(MATRIX_DIM, blocklengths, displs, vectortype, &matrixtype);
        MPI_Type_commit(&matrixtype);

        print_typeinfo(matrixtype, "vector matrixtype");
        print_matrix(matrix, MATRIX_DIM);

        MPI_Send(&matrix[0], 1, matrixtype, 1, tag, MPI_COMM_WORLD);

        MPI_Type_free(&vectortype);
        MPI_Type_free(&matrixtype);
    }
    else
    {
        MPI_Type_contiguous(MATRIX_DIM * MATRIX_DIM, MPI_INT, &matrixtype);
        MPI_Type_commit(&matrixtype);

        MPI_Recv(&matrix[0], 1, matrixtype, 0, tag, MPI_COMM_WORLD, MPI_STATUS_IGNORE);

        print_typeinfo(matrixtype, "contiguous matrixtype");
        print_matrix(matrix, MATRIX_DIM);

        MPI_Type_free(&matrixtype);
    }

    MPI_Finalize();

    return EXIT_SUCCESS;
}

File: Fortran/4_datatypes/solution/b/derived.f90
module print_helpers
    contains

    subroutine print_typeinfo(datatype, typename)
        use mpi
        implicit none
        integer :: datatype
        character*(*) :: typename

        integer ierr, typesize
        integer(KIND=MPI_ADDRESS_KIND) :: lb, extent

        call MPI_TYPE_SIZE(datatype, typesize, ierr)
        call MPI_TYPE_GET_EXTENT(datatype, lb, extent, ierr)

        write (*, fmt="(a,3(a,i0))") typename, ": size=", typesize, " lb=", &
                lb, ", extent=", extent
    end subroutine print_typeinfo

    subroutine print_matrix(matrix, matrix_dim, rank)
        implicit none
        character(len=40) :: fmt
        integer :: matrix_dim, i, rank
        integer, dimension(matrix_dim,matrix_dim) :: matrix

        write (fmt, *) '(i1,a,', matrix_dim, '(" ",i3))'

        do i = 1, matrix_dim
            write (*,fmt) rank, ":", matrix(i,:)
        end do
    end subroutine

end module print_helpers

program datatypes

    use mpi
    use print_helpers
    implicit none

    integer :: i,j
    integer :: ierr, myRank, numProcs, tag = 0
    integer, dimension(MPI_STATUS_SIZE) :: status

    integer, parameter :: matrix_dim = 10
    integer, dimension(matrix_dim, matrix_dim) :: matrix

    integer :: vectortype, matrixtype
    integer :: numblocks, blocklength, stride, intsize
    integer, dimension(matrix_dim) :: blocklengths
    integer(KIND=MPI_ADDRESS_KIND), dimension(matrix_dim) :: displs

    character(len=40) :: fmt
    write (fmt, *) '(i1,a,', matrix_dim, '(" ",i3))'

    call MPI_INIT(ierr)

    call MPI_COMM_RANK(MPI_COMM_WORLD, myRank, ierr)
    call MPI_COMM_SIZE(MPI_COMM_WORLD, numProcs, ierr)

    if (myRank .eq. 0) then
        do i = 1, matrix_dim
            do j = 1, matrix_dim
                matrix(j,i) = (j-1)*100 + i
            end do
        end do
    else
        do i = 1, matrix_dim
            do j = 1, matrix_dim
                matrix(j,i) = -1
            end do
        end do
    end if

    if (myRank .eq. 0) then
        numblocks = matrix_dim
        blocklength = 1
        stride = matrix_dim

        call MPI_TYPE_VECTOR(numblocks, blocklength, stride, &
                MPI_INTEGER, vectortype, ierr)

        call MPI_TYPE_SIZE(MPI_INTEGER, intsize, ierr)
        blocklengths = (/ (1, i=1, matrix_dim) /)
        displs = (/ (i*intsize, i=0, matrix_dim-1) /)
        call MPI_TYPE_CREATE_HINDEXED(matrix_dim, blocklengths, displs, &
                vectortype, matrixtype, ierr)
        call MPI_TYPE_COMMIT(matrixtype, ierr)

        call print_typeinfo(matrixtype, "vector matrixtype")
        call print_matrix(matrix, matrix_dim, myRank);

        call MPI_SEND(matrix, 1, matrixtype, 1, tag, MPI_COMM_WORLD, ierr)

        call MPI_TYPE_FREE(vectortype, ierr)
        call MPI_TYPE_FREE(matrixtype, ierr)
    else if (myRank .eq. 1) then
        call MPI_TYPE_CONTIGUOUS(matrix_dim * matrix_dim, MPI_INTEGER, matrixtype, ierr)
        call MPI_TYPE_COMMIT(matrixtype, ierr)

        call print_typeinfo(matrixtype, "contiguous matrixtype")

        call MPI_RECV(matrix, 1, matrixtype, 0, tag, MPI_COMM_WORLD, status, ierr)
        call print_matrix(matrix, matrix_dim, myRank)

        call MPI_TYPE_FREE(matrixtype, ierr)
    endif

    call MPI_FINALIZE(ierr)

end program datatypes
